New article: "Event Conversion Measurement API" (#4015)
GoogleChrome/web.dev: src/site/content/en/blog/conversion-measurement/index.md

---
title: A more private way to measure ad conversions, the Event Conversion Measurement API
subhead: >

>   A new web API available as an Origin Trial in Chrome 86 measures when an ad click leads to a conversion, without using cross-site identifiers.

Reviewer: ```suggestion
  A new web API available as an origin trial in Chrome 86 measures when an ad click leads to a conversion, without using cross-site identifiers.
```

This isn't capitalized in https://web.dev/origin-trials/

repository. {% endBanner %}

{% Aside %} This API is part of the Privacy Sandbox, a series of proposals to satisfy third-party
use cases without third-party cookies or other cross-site tracking mechanisms. See [Digging into the
Privacy Sandbox](/digging-into-the-privacy-sandbox) for an overview of all the proposals. {%
endAside %}

## Glossary

- **Adtech platforms**: companies that provide software and tools to enable brands or agencies to
  target, deliver, and analyze their digital advertising.
- **Advertisers**: companies paying for advertising.
- **Publishers**: companies that display ads on their websites.
- **Click-through conversion**: conversion that is attributed to an ad click.
- **View-through conversion**: conversion that is attributed to an ad impression (if the user
  doesn't interact with the ad, then later converts).

## Who needs to know about this API: adtech platforms, advertisers, and publishers

- **Adtech platforms** such as **[demand-side
  platforms](https://en.wikipedia.org/wiki/Demand-side_platform)** are likely to be interested in
  using this API to support functionality that currently relies on third-party cookies. If you're

>   working on conversion measurement systems: [try out the demo](#demo), see if you can [register for

Reviewer: ```suggestion
  working on conversion measurement systems: [try out the demo](#demo), [register for
```

  an origin trial](<#register-for-the-origin-trial-(starting-in-chrome-86)>), and [share your
  feedback](#share-your-feedback).
- **Advertisers and publishers relying on custom code for advertising or conversion measurement**
  may similarly be interested in using this API to replace existing techniques.
- **Advertisers and publishers relying on adtech platforms for advertising or conversion
  measurement** don't need to use the API directly, but the [rationale for this
  API](#why-is-this-needed) may be of interest, particularly if you are working with adtech
  platforms that may integrate the API.

> ## API Overview

Reviewer: ```suggestion
## API overview
```
Sentence case is our site-wide convention

address in a privacy-preserving way the use cases that third-party cookies solved. To name a few:

- The Event Conversion Measurement API helps measure click-through conversions for now, and maybe
  view-through conversions in future iterations.
- The [Aggregate Conversion Measurement
  API](https://github.com/WICG/conversion-measurement-api/blob/master/AGGREGATE.md) provides rich
  reports to advertisers by [aggregating conversions from multiple
  clients](https://github.com/WICG/conversion-measurement-api/blob/master/SERVICE.md).
- [Trust Tokens](/trust-tokens/) help combat fraud and distinguish bots from humans.

<figure class="w-figure">
  <img src="./all-apis.jpg" alt="Overview of some privacy sandbox APIs: trust tokens and conversion measurement (aggregate and event)">
</figure>

**How does the Event Conversion Measurement API compare to third-party cookies?**

- It's **purpose-built** to measure conversions, unlike cookies. This in turn can enable browsers to
  apply more enhanced privacy protections.
- It's **more private**: it makes it difficult to recognize a user across two different top-level

>   sites, for example to link publisher-side and advertiser-side user profiles. See how in the [next
>   section](#how-this-api-preserves-user-privacy).

Reviewer: ```suggestion
  sites, for example to link publisher-side and advertiser-side user profiles. See how in the [How
  this API preserves user privacy](#how-this-api-preserves-user-privacy).
```

### A first iteration

This API is at an **early experimental stage**. What's available in Chrome 86 as an origin trial is
the **first iteration** of the API. Things may change substantially in [future
iterations](#use-cases).

### Only clicks

This API only supports **click-through conversion measurement**. View-through conversion measurement
isn't supported yet, because view-through conversions are harder to measure in a truly
privacy-preserving way. This is an active area of work; you can read more about privacy
considerations in the [API
proposal](https://github.com/WICG/conversion-measurement-api#privacy-considerations).

### How it works

<figure class="w-figure">
  <img src="./diagram-overview.jpg" alt="Diagram: overview of the conversion measurement API steps">
</figure>

> This API can be used with two types of `<a>` links used for advertising:
> 
> - Links (`<a>` elements) in a **first-party** context, such as ads on a social network or a search
>   engine results page;
> - Links (`<a>` elements) in a **third-party iframe**, such as on a publisher site that uses a
>   third-party adtech provider.

Reviewer: ```suggestion
This API can be used with two types of links (`<a>` elements) used for advertising:

- Links in a **first-party** context, such as ads on a social network or a search
  engine results page;
- Links in a **third-party iframe**, such as on a publisher site that uses a
  third-party adtech provider.
```
We only need to define links once

> With this API, such outbound links (`<a>` elements) can be configured with attributes that are

Reviewer: ```suggestion
With this API, such outbound links can be configured with attributes that are
```

advertiser's site, it may access click-time or conversion-time data but it can't link them at all.

In this case, user privacy is preserved but advertisers can't optimize their ad spend. This is why
an alternative like the Event Conversion Measurement API is needed.

### Noising of conversion data

The 3 bits gathered at conversion time are **noised**.

For example, in Chrome's implementation, data noising works as follows: 5% of the time, the API
reports a random 3-bit value instead of the actual conversion data.

This protects users from privacy attacks. An actor trying to misuse the data from several
conversions to create an identifier won't have full confidence in the data they receive—making these
types of attacks more complicated.

Note that it's possible to [recover the true conversion
count](<#(optional)-recovering-the-true-conversion-count>).

Summing up click data and conversion data:

<div class="w-table-wrapper">
  <table class="w-table--top-align">
    <thead>
      <tr>
        <th>Data</th>
        <th>Size</th>
        <th>Example</th>
      </tr>
    </thead>
    <tbody>
      <tr>

>         <td>Click data(<code>impressiondata</code> attribute)</td>

Reviewer: ```suggestion
        <td>Click data (<code>impressiondata</code> attribute)</td>
```

        <td>64 bits</td>
        <td>An ad ID or click ID</td>
      </tr>
      <tr>
        <td>Conversion data</td>
        <td>3 bits, noised</td>

>         <td>An integer from 0 to 7 that can map to a conversion type: signup, complete checkout...</td>

Reviewer: ```suggestion
        <td>An integer from 0 to 7 that can map to a conversion type: signup, complete checkout, etc.</td>
```

details on timing.

## Example

{% Banner 'info', 'body' %} To see this in action, try out the
[demo](https://goo.gle/demo-event-level-conversion-measurement-api) ⚡️ and see the corresponding
[code](https://github.com/GoogleChromeLabs/trust-safety-demo/tree/main/conversion-measurement). {%
endBanner %}

Here's how the API records and reports a conversion. Note that this is how a click-to-convert flow
would work with the current API. Future iterations of this API [may be different](#use-cases).

### Ad click (steps 1 to 5)

<figure class="w-figure">
  <img src="./steps-1.jpg" alt="Diagram: ad click and click storage">
</figure>

An `<a>` ad element is loaded on a publisher site by `adtech.example` within an iframe.

The adtech platform developers have configured the `<a>` element with conversion measurement
attributes:

```html
<a
  id="ad"
  impressiondata="776f09351f5809c5"
  conversiondestination="https://advertiser.example"
  reportingorigin="https://adtech.example"
  impressionexpiry="864000000"
  href="https://advertiser.example/shoes07"
  target="_blank"
>
  <img src="/images/shoe.jpg" alt="shoe" />
</a>
```

This code specifies the following:

> {% Aside %}

Reviewer: @maudnals FYI you were missing a huge table here

![image](https://user-images.githubusercontent.com/4713486/95237460-1952c400-07bd-11eb-83ea-3bf071fe5525.png)

> This code specifies the following:
> 
> {% Aside %}

Reviewer: ```suggestion
This code specifies the following:

<div class="w-table-wrapper">
  <table class="w-table--top-align">
    <thead>
      <tr>
        <th>Attribute</th>
        <th>Default value, maximum, minimum</th>
        <th>Example</th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td><code>impressiondata</code> (required): a <b>64-bit</b> identifier to attach to an ad click.</td>
        <td>(no default)</td>
        <td>A dynamically generated click ID  such as a hex-encoded 64-bit integer:
`776f09351f5809c5`</td>
      </tr>
      <tr>
        <td><code>conversiondestination</code> (required): the <b>origin</b> where a conversion is expected for this ad.</td>
        <td>(no default)</td>
        <td><code>https://advertiser.example</code></td>
      </tr>
      <tr>
        <td><code>impressionexpiry</code> (optional): in milliseconds, the cutoff time for when conversions can be attributed to this ad.</td>
        <td>
          <p><code>2592000000</code> = 30 days (in milliseconds)</p>
          <p>Maximum: 30 days (in milliseconds)</p>
          <p>Minimum: 2 days (in milliseconds)</p>
        </td>
        <td>Ten days after click: <code>864000000</code></td>
      </tr>
      <tr>
        <td><code>reportingorigin</code> (optional): the destination for reporting confirmed conversions.</td>
        <td>Top-level origin of the page where the link element is added.</td>
        <td><code>https://adtech.example</code></td>
      </tr>
      <tr>
        <td><code>href</code>: the intended destination of the ad click.</td>
        <td><code>/</code></td>
        <td><code>https://advertiser.example/shoes07</code></td>
      </tr>
    </tbody>
  </table>
</div>

{% Aside %}
Some notes about the example:

- Note that `_target="parent"` is specified here, because in this example the ad is within an iframe.
```

  src="https://adtech.example/conversion?model=shoe07&type=checkout&…"
/>
```

`adtech.example` receives this request, and decides that it qualifies as a conversion. They now need
to request the browser to record a conversion. `adtech.example` compresses all of the conversion
data into 3 bits—an integer between 0 and 7, for example they might map a **Checkout** action to a
conversion value of 2.

`adtech.example` then sends a specific register-conversion redirect to the browser:

```js
const conversionValues = {
  signup: 1,
  checkout: 2,
};

app.get('/conversion', (req, res) => {
  const conversionData = conversionValues[req.query.conversiontype];
  res.redirect(
    302,
    `/.well-known/register-conversion?conversion-data=${conversionData}`,
  );
});
```

{% Aside %} `.well-known` URLs are special URLs. They make it easy for software tools and servers to
discover commonly-needed information or resources for a site—for example, on what page a user can
[change their password](/change-password-url/). Here, `.well-known` is only used so that the browser
recognizes this as a special conversion request. This request is actually cancelled internally by
the browser. {% endAside %}

> The browser receives this request. Upon detecting ".well-known/register-conversion", the browser:

Reviewer: ```suggestion
The browser receives this request. Upon detecting `.well-known/register-conversion`, the browser:
```

Now, the browser knows that it needs to inform the adtech server of this conversion—more
specifically, the browser must inform the `reportingorigin` that is specified in both the `<a>`
element and in the pixel request (`adtech.example`).

To do so, the browser schedules to send a **conversion report**, a blob of data containing the click
data (from the publisher's site) and the conversion data (from the advertiser's). For this example,
the user converted one day after click. So the report is scheduled to be sent on the next day, at
the two-day-after-click mark if the browser is running.

### Sending the report (steps 10 and 11)

<figure class="w-figure">
  <img src="./steps-3.jpg" alt="Diagram: browser sending the report">
</figure>

Once the scheduled time to send the report is reached, the browser sends the **conversion report**:
it sends an HTTP POST to the reporting origin that was specified in the `<a>` element
(`adtech.example`). For example:

`https://adtech.example/.well-known/register-conversion?impression-data=776f09351f5809c5&conversion-data=2&credit=100`

Included as parameters are:

- The data associated with the original ad click (`impression-data`).

> - The data associated with a conversion, [potentially noised](link to above).

Reviewer: ```suggestion
- The data associated with a conversion, [potentially noised](#noising-of-conversion-data).
```

```

### Subsequent conversions and expiry

Later on, the user may convert again—for example by purchasing a tennis racket on
`advertiser.example` to go alongside their shoes. A similar flow takes place:

- The adtech server sends a conversion request to the browser.
- The browser matches this conversion with the ad click, schedules a report, and sends it to the
  adtech server later on.

After `impressionexpiry`, conversions for this ad click stop being counted and the ad click is
deleted from browser storage.

## Use cases

### What is currently supported

- Measure click-through conversions: determine which ad clicks lead to conversions, and access
  coarse information about the conversion.
- Gather data to optimize ad selection, for example by training machine learning models.

### What is not supported yet

The following features aren't supported yet, but may be in future iterations of this API, or in the
[Aggregate Conversion Measurement
API](https://github.com/WICG/conversion-measurement-api/blob/master/AGGREGATE.md):

- View-through conversion measurement.
- [Multiple reporting endpoints](https://github.com/WICG/conversion-measurement-api/issues/29).

> - [Web conversions that started in a native

Reviewer: ```suggestion
- [Web conversions that started in an iOS/Android
```
FYI @maudnals `native` is a "do not use" word now (based on P&E's efforts to stop using words with problematic histories)

  granular purchase values or product categories.

Before these features and more can be supported, **more privacy protections** (noise, fewer bits, or
other limitations) must be added to the API.

Discussion of additional possible features takes place in the open, in the [**Issues** of the API
proposal repository](https://github.com/WICG/conversion-measurement-api/issues).

{% Aside %} Is your use case missing? Do you have feedback on the API? [Share
it](#share-your-feedback). {% endAside %}

### What else may change in future iterations

- This API is at an early, experimental stage. In future iterations, this API may undergo
  substantial changes including but not limited to the ones listed below. Its goal is to measure
  conversions while preserving user privacy, and any change that would help better address this use
  case will be made.
- API and attribute naming may evolve.
- Click data and conversion data may not require encoding.
- The 3-bit limit for conversion data may be increased or decreased.
- The conversion destination may become an eTLD+1; right now, it's an
  [origin](/same-site-same-origin/#origin).

> - [More features may be added] (see above), and **more privacy protections** (noise / fewer bits /

Reviewer: ```suggestion
- [More features may be added](#what-is-not-supported-yet), and **more privacy protections** (noise / fewer bits /
```

  you can register your origin for an [origin trial](/origin-trials/).

### Share your feedback

**Your feedback is crucial**, so that new conversion measurement APIs can support your use cases and
provide a good developer experience.

- To report a bug on the Chrome implementation, [open a
  bug](https://bugs.chromium.org/p/chromium/issues/entry?status=Unconfirmed&components=Internals>ConversionMeasurement&description=Chrome%20Version%3A%20%28copy%20from%20chrome%3A%2F%2Fversion%29%0AOS%3A%20%28e.g.%20Win10%2C%20MacOS%2010.12%2C%20etc...%29%0AURLs%20%28if%20applicable%29%20%3A%0A%0AWhat%20steps%20will%20reproduce%20the%20problem%3F%0A%281%29%0A%282%29%0A%283%29%0A%0AWhat%20is%20the%20expected%20result%3F%0A%0A%0AWhat%20happens%20instead%3F%0A%0AIf%20applicable%2C%20include%20screenshots%2Finfo%20from%20chrome%3A%2F%2Fconversion-internals%20or%20relevant%20devtools%20errors.%0A).
- To share feedback and discuss use cases on the Chrome API, create a new issue or engage in
  existing ones on the [API proposal
  repository](https://github.com/WICG/conversion-measurement-api/issues). Similarly, you can discuss

>   the Webkit/Safari API and its use cases on the [API proposal

Reviewer: ```suggestion
  the WebKit/Safari API and its use cases on the [API proposal
```

  repository](https://github.com/privacycg/private-click-measurement/issues).
- To discuss advertising use cases and exchange views with industry experts: join the [Improving Web
  Advertising Business Group](https://www.w3.org/community/web-adv/). Join the [Privacy Community

>   Group](https://www.w3.org/community/privacycg/) for discussions around the Webkit/Safari API.

Reviewer: ```suggestion
  Group](https://www.w3.org/community/privacycg/) for discussions around the WebKit/Safari API.
```
